src/ir/ir_traversing_visitor_with_result_test.cc

```diff
@@ -0,0 +1,169 @@
+  }
+
+  void FoldResult(const Operation& operation, ResultType child_result, ResultType& result) override {
+    result.insert( result.end(), child_result.begin(), child_result.end() );
+  }
+  void PostVisit(const Operation& operation, ResultType& result) override {
+    if (post_visits_) result.push_back((void*)std::addressof(operation));
+  }
+
+ private:
+  bool pre_visits_;
+  bool post_visits_;
+};
+
+TEST(IRTraversingVisitorTest, TraversesModuleAsExpected) {
+  Module global_module;
+  const Block* block1 =
+      std::addressof(global_module.AddBlock(std::make_unique<Block>()));
+  const Block* block2 =
+      std::addressof(global_module.AddBlock(std::make_unique<Block>()));
+
+  CollectingVisitor preorder_visitor(CollectingVisitor::TraversalType::PRE);
+  ResultType nodes1 = global_module.Accept<CollectingVisitor, ResultType>(preorder_visitor);
+  EXPECT_THAT(
+      nodes1,
+      testing::ElementsAre(std::addressof(global_module), block1, block2));
+  
+  CollectingVisitor postorder_visitor(CollectingVisitor::TraversalType::POST);
+  ResultType nodes2 = global_module.Accept(postorder_visitor);
+  EXPECT_THAT(
+      nodes2,
+      testing::ElementsAre(block1, block2, std::addressof(global_module)));
+
+  CollectingVisitor all_order_visitor(CollectingVisitor::TraversalType::BOTH);
+  ResultType nodes3 = global_module.Accept(all_order_visitor);
+  EXPECT_THAT(
+      nodes3,
+      testing::ElementsAre(std::addressof(global_module), block1, block1,
+                           block2, block2, std::addressof(global_module)));
+}
+
+TEST(IRTraversingVisitorTest, TraversesBlockAsExpected) {
+  auto plus_op = std::make_unique<Operator>("core.plus");
+  auto minus_op = std::make_unique<Operator>("core.minus");
+  BlockBuilder builder;
+  const Operation* plus_op_instance =
+      std::addressof(builder.AddOperation(*plus_op, {}, {}));
+  const Operation* minus_op_instance =
+      std::addressof(builder.AddOperation(*minus_op, {}, {}));
+  auto block = builder.build();
+
+  CollectingVisitor preorder_visitor(CollectingVisitor::TraversalType::PRE);
+  ResultType nodes1 = block->Accept<CollectingVisitor, ResultType>(preorder_visitor);
+  EXPECT_THAT(
+      nodes1,
+      testing::ElementsAre(block.get(), plus_op_instance, minus_op_instance));
+
+  CollectingVisitor postorder_visitor(CollectingVisitor::TraversalType::POST);
+  ResultType nodes2 = block->Accept(postorder_visitor);
+  EXPECT_THAT(
+      nodes2,
+      testing::ElementsAre(plus_op_instance, minus_op_instance, block.get()));
+
+  CollectingVisitor all_order_visitor(CollectingVisitor::TraversalType::BOTH);
+  ResultType nodes3 = block->Accept(all_order_visitor);
+  EXPECT_THAT(
+      nodes3,
+      testing::ElementsAre(block.get(), plus_op_instance, plus_op_instance,
+                           minus_op_instance, minus_op_instance, block.get()));
+}
+
+TEST(IRTraversingVisitorTest, TraversesOperationAsExpected) {
+  auto plus_op = std::make_unique<Operator>("core.plus");
```

Comment:
You can use the IR parser to create the input module: https://github.com/google-research/raksha/blob/main/src/parser/ir/ir_parser_test.cc

Reply:
That's a handy approach, thanks. Mark had previously suggested using the proto text format, but thanks for the options there.

Reply:
> Mark had previously suggested using the proto text format, 

To use the proto text format, won't you need the proto deserializer, which is not yet available?  I was suggesting the IR parser for testing the IRTraversingVisitor in particular.

Reply:
I also have a proto deserializer in this change :)